File: src/services/sigv4.ts
// Firma SigV4 para AWS, sin dependencias.
//
// Bedrock no acepta una API key pelada en un header como OpenAI: cada request
// va firmado con las credenciales (temporales, en el caso del SSO del curso).
// Meter el SDK de AWS solo para esto traería medio árbol de dependencias y
// rompería la premisa del repo, así que el algoritmo va acá. Está fijado por
// AWS y no tiene variantes: es código que se escribe una vez y no se toca.
import { createHash, createHmac } from 'node:crypto';

const ALGO = 'AWS4-HMAC-SHA256';

export interface AwsCreds {
  accessKeyId: string;
  secretAccessKey: string;
  /** Presente solo en credenciales temporales (SSO/STS). */
  sessionToken?: string;
}

export interface RequestFirmado {
  url: string;
  headers: Record<string, string>;
  body: string;
}

function sha256Hex(data: string): string {
  return createHash('sha256').update(data, 'utf8').digest('hex');
}

function hmac(key: Buffer | string, data: string): Buffer {
  return createHmac('sha256', key).update(data, 'utf8').digest();
}

// Clave derivada en cadena: secreto → fecha → región → servicio → aws4_request.
function claveDeFirma(secret: string, fecha: string, region: string, servicio: string): Buffer {
  const kDate = hmac('AWS4' + secret, fecha);
  const kRegion = hmac(kDate, region);
  const kService = hmac(kRegion, servicio);
  return hmac(kService, 'aws4_request');
}

// RFC 3986 reserva estos cuatro, que encodeURIComponent deja pasar.
function encodeRfc3986(s: string): string {
  return s.replace(/[!'()*]/g, (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase());
}

/**
 * Path canónico: el que se firma NO es el que se envía.
 *
 * Para todo servicio que no sea S3, AWS percent-encodea de nuevo un path que ya
 * viene encodeado. Un id de modelo como `amazon.nova-lite-v1:0` viaja en la URL
 * como `...v1%3A0` y se firma como `...v1%253A0`. Verificado contra el
 * canonical request que imprime botocore con `--debug`.
 */
function pathCanonico(pathEnviado: string): string {
  return encodeRfc3986(encodeURIComponent(pathEnviado)).replace(/%2F/g, '/');
}

/**
 * Firma un POST y devuelve la URL, los headers y el body listos para `fetch`.
 *
 * `path` es el path tal como se envía, ya percent-encodeado por quien llama
 * (los ids de modelo de Bedrock traen `:`). La doble codificación que exige la
 * firma se aplica acá adentro.
 */
export function firmarAws(opts: {
  method: string;
  host: string;
  path: string;
  region: string;
  service: string;
  body: string;
  creds: AwsCreds;
  contentType?: string;
  ahora?: Date;
}): RequestFirmado {
  const { method, host, path, region, service, body, creds } = opts;
  const now = opts.ahora ?? new Date();

  // YYYYMMDDTHHMMSSZ
  const amzDate = now.toISOString().replace(/[:-]|\.\d{3}/g, '');
  const fecha = amzDate.slice(0, 8);

  // `host` se firma pero no se manda: lo pone el runtime desde la URL, y
  // mandarlo a mano lo duplicaría.
  const firmados: Record<string, string> = {
    'content-type': opts.contentType ?? 'application/json',
    host,
    'x-amz-date': amzDate,
  };
  if (creds.sessionToken) firmados['x-amz-security-token'] = creds.sessionToken;

  const nombres = Object.keys(firmados).sort();
  const canonicalHeaders = nombres.map((n) => `${n}:${firmados[n].trim()}\n`).join('');
  const signedHeaders = nombres.join(';');

  // method / uri / query / headers / signedHeaders / hash(payload).
  // El join intercala el \n que separa el bloque de headers (que ya termina
  // en \n) de la lista de signedHeaders.
  const canonicalRequest = [method, pathCanonico(path), '', canonicalHeaders, signedHeaders, sha256Hex(body)].join('\n');

  const scope = `${fecha}/${region}/${service}/aws4_request`;
  const stringToSign = [ALGO, amzDate, scope, sha256Hex(canonicalRequest)].join('\n');
  const firma = hmac(claveDeFirma(creds.secretAccessKey, fecha, region, service), stringToSign).toString('hex');

  const { host: _omitido, ...headers } = firmados;
  return {
    url: `https://${host}${path}`,
    headers: {
      ...headers,
      Authorization: `${ALGO} Credential=${creds.accessKeyId}/${scope}, SignedHeaders=${signedHeaders}, Signature=${firma}`,
    },
    body,
  };
}

File: backend/service/sigv4.py
"""Firma SigV4 para AWS, sin dependencias.

Bedrock no acepta una API key pelada en un header: cada request va firmado
con las credenciales (temporales, en el caso del SSO del curso).
Meter el SDK de AWS solo para esto traería medio árbol de dependencias y
rompería la premisa del repo, así que el algoritmo va acá. Está fijado por
AWS y no tiene variantes: es código que se escribe una vez y no se toca.
"""
import hashlib
import hmac as _hmac
from dataclasses import dataclass
from datetime import datetime, timezone
from urllib.parse import quote

ALGO = 'AWS4-HMAC-SHA256'


@dataclass
class AwsCreds:
    access_key_id: str
    secret_access_key: str
    #: Presente solo en credenciales temporales (SSO/STS).
    session_token: str | None = None


@dataclass
class RequestFirmado:
    url: str
    headers: dict[str, str]
    body: str


def _sha256_hex(data: str) -> str:
    return hashlib.sha256(data.encode('utf-8')).hexdigest()


def _hmac_bytes(key: bytes, data: str) -> bytes:
    return _hmac.new(key, data.encode('utf-8'), hashlib.sha256).digest()


def _clave_de_firma(secret: str, fecha: str, region: str, servicio: str) -> bytes:
    """Clave derivada en cadena: secreto → fecha → región → servicio → aws4_request."""
    k_date = _hmac_bytes(('AWS4' + secret).encode('utf-8'), fecha)
    k_region = _hmac_bytes(k_date, region)
    k_service = _hmac_bytes(k_region, servicio)
    return _hmac_bytes(k_service, 'aws4_request')


def _path_canonico(path_enviado: str) -> str:
    """Path canónico: el que se firma NO es el que se envía.

    Para todo servicio que no sea S3, AWS percent-encodea de nuevo un path que ya
    viene encodeado. Un id de modelo como `amazon.nova-lite-v1:0` viaja en la URL
    como `...v1%3A0` y se firma como `...v1%253A0`. Verificado contra el
    canonical request que imprime botocore con `--debug`.

    `quote(safe='')` deja sin tocar exactamente los caracteres no reservados de
    RFC 3986 (A-Za-z0-9 - _ . ~), que es lo que pide la firma; las barras se
    reponen después porque separan segmentos y no se codifican.
    """
    return quote(path_enviado, safe='').replace('%2F', '/')


def firmar_aws(
    *,
    method: str,
    host: str,
    path: str,
    region: str,
    service: str,
    body: str,
    creds: AwsCreds,
    content_type: str = 'application/json',
    ahora: datetime | None = None,
) -> RequestFirmado:
    """Firma un POST y devuelve la URL, los headers y el body listos para el envío.

    `path` es el path tal como se envía, ya percent-encodeado por quien llama
    (los ids de modelo de Bedrock traen `:`). La doble codificación que exige la
    firma se aplica acá adentro.
    """
    now = ahora if ahora is not None else datetime.now(timezone.utc)

    # YYYYMMDDTHHMMSSZ
    amz_date = now.astimezone(timezone.utc).strftime('%Y%m%dT%H%M%SZ')
    fecha = amz_date[:8]

    # `host` se firma pero no se manda: lo pone el runtime desde la URL, y
    # mandarlo a mano lo duplicaría.
    firmados: dict[str, str] = {
        'content-type': content_type,
        'host': host,
        'x-amz-date': amz_date,
    }
    if creds.session_token:
        firmados['x-amz-security-token'] = creds.session_token

    nombres = sorted(firmados)
    canonical_headers = ''.join(f'{n}:{firmados[n].strip()}\n' for n in nombres)
    signed_headers = ';'.join(nombres)

    # method / uri / query / headers / signedHeaders / hash(payload).
    # El join intercala el \n que separa el bloque de headers (que ya termina
    # en \n) de la lista de signedHeaders.
    canonical_request = '\n'.join(
        [method, _path_canonico(path), '', canonical_headers, signed_headers, _sha256_hex(body)])

    scope = f'{fecha}/{region}/{service}/aws4_request'
    string_to_sign = '\n'.join([ALGO, amz_date, scope, _sha256_hex(canonical_request)])
    firma = _hmac_bytes(_clave_de_firma(creds.secret_access_key, fecha, region, service), string_to_sign).hex()

    headers = {n: v for n, v in firmados.items() if n != 'host'}
    headers['Authorization'] = (
        f'{ALGO} Credential={creds.access_key_id}/{scope}, '
        f'SignedHeaders={signed_headers}, Signature={firma}')
    return RequestFirmado(url=f'https://{host}{path}', headers=headers, body=body)
